### DISCUSSION TITLE
Support for: SIMD

### DISCUSSION THREAD

**Discussion by GiorgosXou** [GiorgosXou/NeuralNetworks#16] [Ideas]
Just learned about [SIMD](https://en.wikipedia.org/wiki/Single_instruction,_multiple_data) from a [comment](https://stackoverflow.com/questions/68200504/pangram-in-c-using-functions/68207545?noredirect=1#comment134726522_68207545) about pangrams in stackoverflow. ESP32 I think supports SIMD ... and possibly many others, so why not?

**Comment by StuartIanNaylor** [comment]
Esp32-S3 supports SIMD as it has vector instructions, not so sure about the plain old esp32 though.
https://github.com/espressif/esp-dl
The benchmarks are for the esp-dl that have been optimised with vector instructions.
https://docs.espressif.com/projects/esp-dsp/en/latest/esp32/esp-dsp-benchmarks.html

I think cadence holds some IP as the layers and functions avail are limited such as no LSTM or Leaky Relu...

Would be great to have a full SIMD support on this microcontroller.

**Reply by GiorgosXou** [reply]
Interesting, seems as if only `ESP-32 S3` supports SIMD instructions and right now I only have the `C3`. Thankfully though, I can test\play around with the `ansi` versions of the functions. Thanks for bringing this to my attention, I'll see what I can do.

**Reply by StuartIanNaylor** [reply]
Likely will work the same but with an S3 up to x10 faster that they got with some of there benches.
I am not sure why with tflite4micro why recurrent layers are missing and likely because I haven't a clue at that level.
GRU & LSTM are sadly missed as quite a few small models use them but the S3 is quite a boost with ML and not much more https://www.lilygo.cc/products/t7-s3

**Reply by GiorgosXou** [reply]
@StuartIanNaylor new version, I just added a [basic SIMD optimization](https://github.com/GiorgosXou/NeuralNetworks/blob/5cd31c9a29853899c36b5ca7d0d8cf5e9cb3422e/src/NeuralNetwork.h#L1964-L1967), improving feedforward for ESP32-S3 aproximatelly from `O(n^3)` to `O(n^2)`.

**Reply by oursland** [reply]
ESP32-S3 does not have SIMD floating point functionality.  If you look at the [source for dotprod](https://github.com/espressif/esp-dsp/blob/3c12d059f96ec95ed00769fb32d41ca23e20cdf1/modules/dotprod/float/dsps_dotprod_f32_aes3.S#L68), it has special instructions to load 128-bits at a time (4 floats) (`EE.LDF.128.IP`) but the actual operations are scalar multiplies.

It *does* however have SIMD functionality for small integers.  In the [signed char version of dotprod](https://github.com/espressif/esp-dsp/blame/3c12d059f96ec95ed00769fb32d41ca23e20cdf1/modules/dotprod/fixed/dspi_dotprod_off_s8_aes3.S#L377) you will see both 128 bit loads and saves, but also vector operations (`ee.vmul` prefix).  These instructions are also used heavily in Espressif's NN library: [esp-nn](https://github.com/espressif/esp-nn/tree/master).

**Reply by GiorgosXou** [reply]
> ESP32-S3 does not have SIMD floating point functionality.

@oursland **SIMD stands for Single Instruction, Multiple Data;** one instruction operates on multiple data elements in parallel. The ESP32-S3 dot-product instructions:

- Take a vector of 4 floats from memory *(`EE.LDF.128.IP`)*
- Take another 4 floats *(from a register or another memory location)*
- Perform 4 multiply-accumulate operations in parallel
- Accumulate the partial results appropriately

**This is a SIMD operation.**

**Reply by GiorgosXou** [reply]
**However,** thanks for pointing out the the [signed char version of dotprod](https://github.com/espressif/esp-dsp/blame/3c12d059f96ec95ed00769fb32d41ca23e20cdf1/modules/dotprod/fixed/dspi_dotprod_off_s8_aes3.S#L377) and the Espressif's NN library: [esp-nn](https://github.com/espressif/esp-nn/tree/master). I might take a deeper look into them, they seem interesting.

**Reply by oursland** [reply]
See this in the code I referenced:

```assembly
		madd.s	f0, f4, f8		// f0 = X11*Y11
		madd.s	f1, f5, f9		// f1 = X12*Y11
		madd.s	f2, f6, f10		// f2 = X13*Y11
		madd.s	f3, f7, f11		// f3 = X14*Y11
```

That's 4 scalar operations, not SIMD.

**Reply by GiorgosXou** [reply]
@oursland so, what is the correct terminology here? Loop unrolling with multi-accumulators?

**Reply by oursland** [reply]
Yes, however the code is not too dissimilar to a naively written C loop.  Here the performance gained is entirely from the improved loading throughput.

All of the SIMD functionality on the ESP32-S3 are in the s16 (int16_t) and s8 (int8_t) types and require hand written assembly.  Effectively, this means that it may accelerate some dense networks that have been quantized down to one of the supported types.

**Reply by GiorgosXou** [reply]
**I was wrong.** Thank you @oursland  for pointing this out. I'll fix this **asap**, because I don't want to mislead people. **Last question**, is "ESP32-S3 DSP-accelerated dot-prod." a valid description?

**Reply by oursland** [reply]
Yeah, that's an appropriate description.

If you choose to extend this to work with quantized values, there is a path for running SIMD on the ESP32-S3.

**Reply by GiorgosXou** [reply]
Done https://github.com/GiorgosXou/NeuralNetworks/commit/e3bdead65a01382750efac100aa9666465a6a1ad. Now I'll release a new version and try to correct all my reddit posts to be accurate. Again, **thank you** for taking the time to correct me.

**Reply by GiorgosXou** [reply]
Now that I'm done with [this stupid TODO-implementation-note](https://github.com/GiorgosXou/NeuralNetworks/commit/965c08d189880ef68792a0b2d9de152747eaee18), I'm pretty sure that at least int-quantized NNs will get SIMD support at some point

**Comment by GiorgosXou** [comment]
here's some references and a minimal example for anyone confushed by the limited documentation:
<details>
  <summary>A basic example of code using "dsps_add_f32"</summary>
  
```c++
#include <esp_dsp.h>
#define VECTOR_SIZE 16


void printv(float * v){
  Serial.print("Vector: ");
  for (int i = 0; i < VECTOR_SIZE; i++) {
    Serial.print(v[i]);
    Serial.print(" ");
  }  
  Serial.println();
}


void setup() {
  Serial.begin(115200);

  float a[VECTOR_SIZE];
  float b[VECTOR_SIZE];
  float result[VECTOR_SIZE];

  // Initialize the arrays with some data
  for (int i = 0; i < VECTOR_SIZE; i++) {
    a[i] = i;
    b[i] = VECTOR_SIZE - i;
  }
  printv(a);
  printv(b);
  
  dsps_add_f32(a, b, result, VECTOR_SIZE, 1, 1, 1); // Works just fine on any ESP-32 because a macro translates it to dsps_add_f32_ansi if it doesn't support SIMD instructions
  //dsps_add_f32_ae32(a, b, result, VECTOR_SIZE, 1, 1, 1); // Compiles only for ESP-32-S3 at the momment
  printv(result);
}

void loop() {}
```
</details>

`dsps_add_f32` defined as [`ansi`](https://github.com/espressif/esp-dsp/blob/master/modules/math/add/float/dsps_add_f32_ansi.c) or [`ae32`](https://github.com/espressif/esp-dsp/blob/master/modules/math/add/float/dsps_add_f32_ae32.S) based on [support](https://github.com/espressif/esp-dsp/blob/74e23cd649c40d7835c9de3fc02f69bc838508e9/modules/math/add/include/dsps_add.h#L59-L65)

**Comment by GiorgosXou** [comment]
Partially done *(at least for ESPs)*: https://github.com/GiorgosXou/NeuralNetworks/blob/5cd31c9a29853899c36b5ca7d0d8cf5e9cb3422e/src/NeuralNetwork.h#L1964-L1967
